Transfers: further rework get transfer requests. #4499. Closes #4516.

rcarpa opened the pull request:
Continue rework of transfers internal code. This PR has multiple functional changes: 

- there is no early binding when iterating over sources. All candidate transfers, including intermediate hops, are generated between possible sources and destination, followed by a merge operation. This should have a higher computational and memory cost: a small dict will be generated in memory for each hop, instead of one dict per transfer request. Also, generating this dict takes cpu time, but we mostly speak about string operations. The positive side is that this change should fix multiple issues with previous code. For example: impossibility to find the best source for multipath transfers; impossibility to correctly chose between mixed disk and tape sources; inconsistent transfer definition generated for intermediate jumps in multipaths; etc

- source and destination generation for multipath transfers now use common code path with singlepath ones. The common code  is capable to handle more cases than needed for multipath transfers. For example: it can correctly handle source and destination schemes which are not used for multipath today. However, it allows to reduce the amount of duplicate code.

- the OIDC support is activated after creating multihop transfers. Meaning it could be activated for each intermediate hop separately.

The PR is split into multiple commits for easier review, but it's probably better to squash them on merge. Some of them cannot be safely reverted individually. 

Resolves #4495

bziemons committed:
refactor code of list transfer requests

Use a class rather than a 20+ long tuple for returning query.all()
result list.

This commit leaves a broken preparer, but it is fixed in the next
commit.

Commit:
make single/multihop prioritization work

Add additional tests

Use the computed distance from dijkstra computation during the
source comparison

Commit:
extract transfer preparation from main function

Perform grouping by request_id upstream:
Both current consumers of the data need to group the
requests by their id. Perform this action in the called
function rather than in both consumers.

Commit:
first prepare transfer definitions, then merge sources

will allow to easier handle the special cases of
multihop sources and tape sources later

Commit:
split function into two and reuse part in multihop code

the used function handles more cases than the
substituted code, but seems harmless.

Commit:
extract the context class to main file scope

It is now used in multiple functions. This also allows to
unclog get_transfer_requests_and_source_replicas a bit

Commit:
do not deepcopy transfers

we are only interested to have a copy of the dictionary keys

Commit:
add identation level

To make the diff in next commit mor readable.
Serves no other purpose

Commit:
first generate multihop transfer definitions, than create requests

Until now, a rather cryptic technique was used for multipath transfers:
a transfer definition was generated as if there existed a single hop
path between the source and destination. After that, for each other
hop, a new replica and request was created and a new transfer added
to the transfers dict. The initial transfer definition was modified
to become the definiton of the last hop.
Many artificial limitations (probably even bugs) were introduced by
this behavior. Some difficult to spot:
- the generation of intermediate hop transfers didn't correctly set
the globus endpoint ids; checksum verification; external host;
account; overwrite parameter;
- the last hop transfer definition had incorrect schemes; checksum
verification;

I believe that the limited feature set for intermediate transfers
wasn't an issue because those features are not used by anybody, or
things just worked by chance. For example, incorrect schemes for last
transfer, or incorrect checksum verification type for intermediate
hops just occurred to work because the hops supported those parameters.

Change the previous behaviour. Now directly generate each hop transfer
definition at the same place. This way, the code is mutualized and
intermediate transfers should support the exact same feature set as
single-hop ones. This way, some particularities of the last hop are
also much more visible (the only one to correctly set the source
ranking). I don't change this behavior, but it's probably worth
challenging it in a further commit.

Because intermediate TAPE hops are now "theoretically" possible (the
tape RSE must have the used_for_multihop=True attribute), add an
additional protection against this case.

rcarpa commented:
The new integration tests I added recently showed some regressions. Working on fixing them

Commit:
remove un-needed identations

bziemons commented on lib/rucio/core/transfer.py:
```diff
@@ -713,376 +1057,203 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
+                                                                        rses=rses,
+                                                                        request_state=RequestState.QUEUED,
+                                                                        transfertool=transfertool,
+                                                                        session=session)
```
imo better:
```py
    request_with_sources = __list_transfer_requests_and_source_replicas(
        total_workers=total_workers,
        worker_number=worker_number,
        limit=limit,
        activity=activity,
        older_than=older_than,
        rses=rses,
        request_state=RequestState.QUEUED,
        transfertool=transfertool,
        session=session,
    )
```
see Black formatter.

rcarpa replied:
I agree that it's better this way

bziemons commented on lib/rucio/core/transfer.py:
```diff
@@ -1278,19 +1288,28 @@ def protocol(self, rse_id, scheme, operation):
     :param transfertool:     The transfer tool as specified in rucio.cfg.
     :param session:          Database session to use.
-    :returns:                List.
+    :returns:                List of RequestWithSource namedtuple.
```
adjust comment

rcarpa replied:
done

bziemons commented on lib/rucio/core/request.py:
```diff
@@ -1592,48 +1590,48 @@ def __throttler_request_state(activity, source_rse_id, dest_rse_id, session: "Op
     """
     Reduces the passed request sources tuples by using the sort-reduce functions,
     returning the first of the remaining items or a simple
     (request_id, RequestState.NO_SOURCES) tuple, if no sources were found.
```
doc needs adjustment
```py
    """
    Reduces the passed requests & sources objects by using the sort-reduce
    functions, yielding the best RequestResult object or a RequestAndState
    object. If all sources were filtered, a RequestAndState object with
    RequestState.NO_SOURCES is yielded.
    """
```

rcarpa replied:
You even provided the entire correct docstring ! Thanks. Updated.

bziemons commented on lib/rucio/core/transfer.py:
```diff
@@ -1382,14 +1398,25 @@ def __list_transfer_requests_and_source_replicas(total_workers=0, worker_number=
+        if not request:
+            request = RequestWithSources(id=request_id, rule_id=rule_id, scope=scope, name=name, md5=md5, adler32=adler32, byte_count=byte_count,
+                                         activity=activity, attributes=attributes, previous_attempt_id=previous_attempt_id, dest_rse_id=dest_rse_id,
+                                         account=account, retry_count=retry_count)
```
is there any specific reason not to use a `namedtuple` and get rid of great unwrapping in the for-loop?

rcarpa replied:
Namedtuples are immutable. There were a couple of places where having a fully mutable class allowed to simplify the logic. For example: to sanitize the source_ranking and set it to 0 in the case when it is returned as None. I'm however not very sure the benefit is really justified right now. Named tuples could do the trick quite well. It ads complexity in some places, but makes life easier in others.... However, I'm thinking of (maybe) bringing some logic into this class in the future for a more object-oriented design. But I'm far from that design for now, and I'll maybe revert back to named tuples if I find it better suited latter.

Commit:
add identation level

mlassnig commented:
@rcarpa LGTM, just this here:

Conflicting files:
`lib/rucio/core/transfer.py`

rcarpa commented:
> @rcarpa LGTM, just this here:
> 
> Conflicting files:
> `lib/rucio/core/transfer.py`

Rebased. The conflict was with one of my recent hotfixes. The hotfix doesn't apply to this PR, as the affected code was completely reworked and removed in this PR.

bari12 commented on lib/rucio/core/request.py:
```diff
@@ -1544,38 +1550,30 @@ def list_requests(src_rse_ids, dst_rse_ids, states=[RequestState.WAITING], sessi
-
 @transactional_session
-def preparer_update_requests(source_iter: "Iterable[Sequence]", session: "Optional[Session]" = None) -> int:
+def preparer_update_requests(source_iter: "Iterable[RequestResultOrState]", session: "Optional[Session]" = None) -> int:
```
Please add some documentation what the method is supposed to do. (I know, we didn't have this before either, but let's make it better readable for the next generation 😄 ) - Maybe this is more for @bziemons actually.

rcarpa replied:
Done with a quite generic docstring. Didn't have much inspiration for this one. The function does a couple of different things which are difficult to summarize into a docstring.

bziemons replied:
I mean, it updates the requests in the database based on its input essentially.. I still think the function name is descriptive enough

bari12 commented on lib/rucio/core/request.py:
```diff
@@ -1615,48 +1613,49 @@ def __throttler_request_state(activity, source_rse_id, dest_rse_id, session: "Op
 
 
-def reduce_requests(req_sources: "List[Tuple]", sort_reduce_funcs: "List[ReduceFunction]", logger: "Callable") -> "RowIterator":
+def reduce_requests(
```
Just out of personal preference, break the parameters in individual lines instead?

rcarpa replied:
done

bari12 commented on lib/rucio/core/request.py:
```diff
@@ -1675,37 +1674,38 @@ def get_supported_transfertools(rse_id: str, session=None) -> "Set[str]":
 
 
 def get_transfertool_filter(
```
See above

rcarpa replied:
not sure what to do with this one. There is a single argument on a separate line. The function name is expressive enough that I don't have good ideas for a docstring.

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -119,6 +119,83 @@
 WEBDAV_TRANSFER_MODE = config_get('conveyor', 'webdav_transfer_mode', False, None)
 
 
+class TransferSource:
```
`@dataclass` ?

rcarpa replied:
Would love it. But I think it's a python 3.7+ feature.

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
     return dest_url
 
 
+def __prepare_source_dest_config(ctx, request_id, src_rse_id, dest_rse_id, source_scheme, dest_scheme, dict_attributes, activity,
```
Add a short description what this does.

rcarpa replied:
merged this function into the calling context. I feel it is better this way.

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+
+
+@transactional_session
+def __prepare_transfer_definition(ctx, rws, source, computed_distance, dict_attributes, transfertool, retry_other_fts, list_hops, activity,
```
Also add quick description

rcarpa replied:
done

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+    return transfer_path
+
+
+def __merge_or_discard_tranfer_definitions(candidate_paths):
```
Description

rcarpa replied:
done

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -723,379 +1067,205 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
-                        if req_id in reqs_no_source:
-                            reqs_no_source.remove(req_id)
-                        continue
+            # rse_name will be None if rse is staging area
```
Where is this enforced?

rcarpa replied:
Didn't think this through. The comment was there before the refactoring. Do you feel like it is wrong ?

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+
+
 @transactional_session
 def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0, limit=None, activity=None, older_than=None, rses=None, schemes=None,
```
Add a general description here about the further workflow. Nothing huge, just to ilustrate the steps being done to build the chosen transfers.

rcarpa replied:
done. added a short workflow description at the end of the docstring

bari12 reviewed (changes requested):
Thanks for the excellent PR Radu and sorry for the delay, just a few comments. (The preparer ones might actually be for @bziemons )
I went over it, back and forth for some time now, and I think it is correct. But it's a big PR, so be ready for hotfixes anyway 😄 
I especially like the new context classes.
Please still add a bit more documentation to the internal methods. No need to do it for each parameter, but just a rough description what the method does, to make it easier for others to understand. 
Also add some general description of the workflow. Although it is easier now, it is still not easy to understand. I think on the entry point of the `get_transfer_requests_and_source_replicas` you should add something.
Thanks a lot.

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+        transfer_dst_type = "TAPE"
+
+    # Get destination protocol
+    dest_protocol = ctx.protocol(dest_rse_id, dest_scheme, 'write')
```
write -> third_party_copy

rcarpa replied:
done

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+                                 scope, name, file_path, retry_count, reqs_only_tape_source, reqs_no_source, bring_online_local):
+
+    # Get source protocol
+    source_protocol = ctx.protocol(src_rse_id, source_scheme, 'read')
```
read -> third_party_copy 
Was this read before in the old code?

rcarpa replied:
it was . it's just that ` ctx.protocol(src_rse_id, source_scheme, 'read')` doesn't correctly use this parameter . This issue was mentioned by @cserf in the previous refactoring. However, the previous refactoring tried to keep the same behavior as before. This time it's the moment to fix it. I'll make the change.

rcarpa replied:
done

bari12 commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+            # do not add multihop sources
+            continue
+
+        if not set(best_candidate[0]['schemes']).intersection(candidate[0]['schemes']):
```
I have to think a bit more about this, but I am not 100% sure if this comparison is correct.
Because the `schemes` is already the compatible schemes of a specific chosen destination_scheme. But here you are iterating candidates (src-dest pairs) whose compatible_schemes is based on a different selected destination_scheme. I am not 100% sure if this is correct.

Commit:
add docstrings

Commit:
merge __prepare_source_dest_config back into calling context

I initially made it a a separate function because it was used in
multiple places, but it's not the case anymore.

It makes the __prepare_transfer_definition function bigger, but
the flow is very linear and seems much better like that. Having the
logic in a separate function requires passing-through almost all
local variables from the calling context.

Commit:
correctly pass operation on protocol creation

The previous behavior was strange, but I kept it unchanged during
previous refactoring. We were having two separate local "cache"
keys per rse and per protocol : one for inbound transfers and one
for outbound transfers. In both cases, the "third_party_copy" operation
was used.

Fix this counter-intuitive behavior. This also should result in less
"protocol" objects being created. The same protocol object will be
used to compute inbound and outbound pfns for transfer as long as
the scheme is the same.

Commit:
split function arguments into one per line

rcarpa reviewed (commented):
Took into consideration all comments except one. Will have to dig more attentively into the multisource scheme merge. Will add a separate commit for that comment latter.

Commit:
build multisource transfers when possible

in case the best_path is a single-hop DISK RSE, re-iterate over
directly connected sources while enforcing scheme compatibility.

include the hop_distance and cumulated_distance in the get_hops
return value.

As we iterate over sources multiple times, create a list of filtered
sources before the first iteration.

Commit:
rework next state construction for requests without sources

Instead of constructing lists which are updated in multiple places,
just update them once, when all the needed information is available.

Make use of the allow_tape_source attribute on the request.
It was set to a "True" constant before.

Switch to using sets, to simplify handling.

Commit:
retrieve multihop config only once per run

Commit:
swap variable names: bring_online and bring_online_local

Commit:
continue instead of break on exception in source handling

try other sources. There is no reason to skip them

Commit:
remove code which will always be skipped

__list_transfer_requests_and_source_replicas does the same filtering
and is the source of data which is analyzed in this context.

rcarpa commented:
I rebased and added a couple of simple cleanup commits. No changes to previous commits.

Commit:
make attributes a property of RequestWithSources class

Commit:
do not pass activity to __prepare_transfer_definition

it is already present on the rws object.

cserf commented on lib/rucio/core/transfer.py:
```diff
@@ -721,381 +1058,239 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
+        filtered_sources = filter(lambda s: s.rse_id not in unavailable_read_rse_ids, filtered_sources)
+        # Ignore tape sources if they are not desired
+        filtered_sources = list(filtered_sources)
+        had_tape_sources = len(filtered_sources) > 0
```
I don't get it. From the name I understand what it should be (files that had tape sources), but I cannot relate it to `filtered_sources`

rcarpa replied:
check the usage of `had_tape_sources` bellow and it will become clear. It's used like "if not filtered_sources and had_tape_sources .... then". Do you have a better idea on how to achieve this goal ?

Commit:
more object-oriented-programming in transfers

Group information associated to an RSE into a RseData class.
Use it instead of rse_id where convenient.

Introduce DirectTransferDefinition class, which is a wrapper
to the existing dict-based definition. To be used in latter
commits.

Commit:
remove unused transfer data

- dest_scheme_priority

Commit:
move source and destination url computation as class members

This is a step towards removing any transfertool-specific code from
transfer core. The reason it was there in the first place is because
we need RSE (attribute) information; and source (ranking) information
to take some transfertool-specific decisions. Until now, this structured
data was lost at the interface when leaving transfer core. This
"transfer dict" interface was heavily abused to return temporary data
which was then removed from the dict before sending to transfertool.

Rework everything to return more complex objects out of transfer core
these structured objects could latter be used to generate directly
the needed transfertool api call without going via an intermediary
dictionary.

Also remove the un-needed source sorting logic in submitter. They are
already sorted upsteam in __find_compatible_direct_sources.
Moreover, this logic conflicted with the new behavior, because "legacy"
sources cannot be set, they are computed dynamically.

cserf commented on lib/rucio/core/transfer.py:
```diff
@@ -637,6 +719,44 @@ def get_dsn(scope, name, dsn):
+                rse_settings_src=__load_rse_settings(rse_id=source.rse_id, session=session),
+                rse_settings_dest=__load_rse_settings(rse_id=dest_rse_id, session=session),
+                operation_src='third_party_copy',
+                operation_dest='third_party_copy',
```
Not to forget. We should use `third_party_copy_write`  and `third_party_copy_read` at some point

cserf commented on lib/rucio/core/transfer.py:
```diff
@@ -702,6 +822,223 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+        # on equal type, prefer lower distance_ranking
+        # on equal distance, prefer single hop
+        return (
+            - transfer_path[-1]['sources'][0][3],
```
So if there are 2 different transfer paths:
1 - A -> dest where A hosts the file but has lower ranking than B
2 - C -> B -> dest where B doesn't host the file but has higher ranking than A
In that case the multihop will be chosen ?

rcarpa replied:
yes, but source_ranking is only decreased at failed transfers. Meaning that A was tried in the past, but failed. So this current transfer is probably result of re-scheduling the failed one.

cserf commented on lib/rucio/core/transfer.py:
```diff
@@ -721,381 +1058,239 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
+                             ignore_availability=False,
+                             dataset_meta=None,
+                             session=session)
+            except Exception as error:
```
In case of exception, the session is rollback if I'm not wrong

rcarpa replied:
@bari12 , any opinion ? It's not a behavior which changed. It looks like new code because github detects the diff wrongly. But this behavior is old (doesn't mean it's correct though)

bari12 replied:
This is fine. It is the same as before. It actually checks if there already is a replica in `add_replicas`. For the case that it did not detect it at that point, but due to another session adding one exactly at this time, this will fire an integrity-constraint error, and the session will be rolled back. But also that should be fine.

rcarpa commented:
As agreed with @bari12 , added 5 additional commits. They fit into this refactoring conceptually; shouldn't introduce behavioral changes; but are important for latter work on generic transfertool handling. So bring them into this already big PR. Had to rebase due to conflict with https://github.com/rucio/rucio/pull/4651/.
